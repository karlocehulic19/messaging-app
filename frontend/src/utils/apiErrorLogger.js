export default async function apiErrorLogger(error) {
  let resJSON;
  if (
    error.response &&
    error.response.headers.get("Content-Type").includes("application/json")
  )
    resJSON = await error.response.json();
  if (resJSON) console.error(resJSON);
  console.error(error);

  return resJSON;
}
