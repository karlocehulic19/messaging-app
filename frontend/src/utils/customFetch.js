import { config } from "../Constants";
import { merge } from "lodash";

export class ResponseError extends Error {
  constructor(message, response) {
    super(message);
    this.response = response;
  }
}

function getDefaultOptions() {
  const defaultOptions = { headers: { "Content-Type": "application/json" } };
  const token = localStorage.getItem("site");

  if (token) {
    defaultOptions.headers.Authorization = `Bearer ${token}`;
  }

  return defaultOptions;
}

export default async function customFetch(url, options) {
  const response = await fetch(
    `${config.url.BACKEND_URL}${url}`,
    merge(getDefaultOptions(), options)
  );
  if (!response.ok) {
    throw new ResponseError("Bad request error", response);
  }

  return response;
}
