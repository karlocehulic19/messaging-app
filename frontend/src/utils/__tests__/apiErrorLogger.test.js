import apiErrorLogger from "../apiErrorLogger";
import { describe, vi, expect, beforeEach } from "vitest";
import { ResponseError } from "../customFetch";

const consoleErrorSpy = vi
  .spyOn(console, "error")
  .mockImplementation(() => undefined);

beforeEach(() => {
  vi.clearAllMocks();
});

describe("apiErrorLogger()", () => {
  it("Logs request.json() if request.json() is present", async () => {
    const testingResponse = new Response(JSON.stringify({ error: "message" }), {
      headers: {
        "Content-Type": "application/json",
      },
    });
    const resErr = new ResponseError("Bad request", testingResponse);

    await apiErrorLogger(resErr);

    expect(consoleErrorSpy.mock.calls[0][0]).toMatchSnapshot();
    expect(consoleErrorSpy.mock.calls[1][0]).toMatchSnapshot();
  });

  it("Logs only error when type of error isn't response error", async () => {
    const testingError = new Error("Ups! Something went wrong.");

    await apiErrorLogger(testingError);

    expect(consoleErrorSpy).toBeCalledTimes(1);
    expect(consoleErrorSpy).toBeCalledWith(testingError);
  });

  it("Logs only error message when response isn't containing json", async () => {
    const testingResponse = new Response(
      "Error text that shouldn't be standard"
    );

    console.log(testingResponse);

    const resErr = new ResponseError("Bad request", testingResponse);

    await apiErrorLogger(resErr);

    expect(consoleErrorSpy).toBeCalledTimes(1);
    expect(consoleErrorSpy).toBeCalledWith(resErr);
  });
});
