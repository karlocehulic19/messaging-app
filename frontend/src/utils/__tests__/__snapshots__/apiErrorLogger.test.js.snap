// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`apiErrorLogger() > Logs request.json() if request.json() is present 1`] = `
{
  "error": "message",
}
`;

exports[`apiErrorLogger() > Logs request.json() if request.json() is present 2`] = `[Error: Bad request]`;
